### PR TITLE
Express notes api - Jeff Kwon

### PR DESCRIPTION
**Pull request by Jykwon24**
https://recordit.co/p9eF2EIBAb

### CODE REVIEW

File: express-notes-api/index.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+const jsonMethod = express.json();
+app.use(jsonMethod);
+
+const data = require('./data.json');
+const dataArray = [];
+
+app.get('/api/notes', (req, res) => {
+  for (const key in data.notes) {
+    if (!data.notes[key]) {
+      res.status(200).json(dataArray);
```
Comment (TimelikeClosure) [inline express-notes-api/index.js]: When is `!data.notes[key]` going to be `true` here?

File: express-notes-api/index.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    } else {
+      dataArray.push(data.notes[key]);
+    }
+  }
+  res.status(200).json(dataArray);
+});
```
Comment (TimelikeClosure) [inline express-notes-api/index.js]: Performing a `GET /api/notes` multiple times in a row should always produce the same output.

File: express-notes-api/index.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+app.get('/api/notes/:id', (req, res) => {
+  const iD = req.params.id;
+  if (isNaN(parseInt(iD))) {
+    res.status(400).send({ error: 'id must be a positive integer' });
```
Comment (TimelikeClosure) [inline express-notes-api/index.js]: A `400 Bad Request` should also trigger if the id is:
- A non-positive number
- A non-integer number

**Review by TimelikeClosure — CHANGES_REQUESTED** [review]

File: express-notes-api/index.js (hunk 1 above)
Comment (Jykwon24) [reply express-notes-api/index.js]: If the data.json file has no properties in data.notes?

Comment (TimelikeClosure) [reply express-notes-api/index.js]: Even if `data.notes` contains an empty object, that condition will be `false`, meaning `res.status(200).json(dataArray);` will never run:

---

image

File: express-notes-api/index.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      dataArray.push(data.notes[key]);
+    }
+  }
+  res.status(200).json(dataArray);
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  const iD = req.params.id;
+  if ((Math.sign(iD) === -1) || (!Number.isInteger(parseInt(iD)))) {
+    res.status(400).send({ error: 'id must be a positive integer' });
```
Comment (TimelikeClosure) [inline express-notes-api/index.js]: Requests like `GET /api/notes/45.123` (not an integer) and `GET /api/notes/0` (not positive) should also result in a `400 Bad Request` response. Test requests like these to verify they produce the desired output.

**Review by TimelikeClosure — CHANGES_REQUESTED** [review]

File: express-notes-api/index.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+app.get('/api/notes/:id', (req, res) => {
+  const iD = req.params.id;
+  if (data.notes[iD]) {
+    if ((Math.sign(iD) === 1) && Number.isInteger(parseInt(iD)) && iD !== '0') {
+      res.status(200).send(data.notes[iD]);
+    }
+    res.status(404).send({ error: `cannot find note with id ${iD}` });
+  }
+  res.status(400).send({ error: 'id must be a positive integer' });
```
Comment (TimelikeClosure) [inline express-notes-api/index.js]: Calling `res.send(...)` doesn't end the middleware function, so it's possible that `send` will be called three times on one request. Please make sure there is exactly one response sent for any given request.

File: express-notes-api/index.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        res.status(201).send(postBody);
+      }
+    });
+  }
+});
+
+app.delete('/api/notes/:id', (req, res) => {
+  const iD = req.params.id;
+  if (isNaN(parseInt(iD))) {
+    res.status(400).send({ error: 'id must be a positive integer' });
```
Comment (TimelikeClosure) [inline express-notes-api/index.js]: Like the `GET` middleware further up, any `id` which isn't a positive integer should trigger a `400 Bad Request` response: https://en.wikipedia.org/wiki/Natural_number

**Review by TimelikeClosure — CHANGES_REQUESTED** [review]

File: express-notes-api/index.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  } else {
+    res.status(400).send({ error: 'id must be a positive integer' });
+  }
+});
+
+app.put('/api/notes/:id', (req, res) => {
+  const iD = req.params.id;
+  const putBody = req.body;
+  if (isNaN(parseInt(iD))) {
+    res.status(400).send({ error: 'id must be a postive integer' });
```
Comment (TimelikeClosure) [inline express-notes-api/index.js]: This one should enforce positive integers as well.

**Review by TimelikeClosure — CHANGES_REQUESTED** [review]
@Jykwon24 Almost there, just missed one validation.

**Review by TimelikeClosure — APPROVED** [review]

**Pull request merged**